File: day2/solution.kt
package part1

import java.io.File

fun main() {
    println(solution())
}

fun String.parseMove(): Pair<Int, Int> =
        this.split(" ").let {
            val value = it[1].toInt()
            when (it.first()) {
                "forward" -> Pair(0, value)
                "down" -> Pair(value, 0)
                else -> Pair(-value, 0)
            }
        }

fun solution(): Int {
    return File("../input.txt")
            .readLines()
            .fold(Pair(0, 0)) { (x, y), it ->
                val (a, b) = it.parseMove()
                Pair(x + a, y + b)
            }
            .let { (f, s) -> f * s }
}

File: day2/solution.clj
(ns playground)

(require '[clojure.java.io :as io])
(require '[clojure.string :as string])


(defn parse [x y]
  (case x
    "forward" (list 0 (Integer/parseInt y))
    "down" (list (Integer/parseInt y) 0)
    (list (- (Integer/parseInt y)) 0)))

(defn parseMove [str]
  (apply parse (string/split str #" ")))

(defn sumTuples [acc next]
  (list (+ (first acc) (first next)) (+ (second acc) (second next))))

(defn solution []
  (with-open [rdr (io/reader "./input.txt")]
    (->> (line-seq rdr)
         (map parseMove)
         (reduce sumTuples)
         (apply *))))


(print (solution))